### ISSUE TITLE
Unassigned variable not reported

### ISSUE THREAD

**Issue by coderangerdan** [Microsoft/TypeScript#15435]
**TypeScript Version:**  2.3.1

**Code**

```ts
var channelCombo = <HTMLSelectElement>document.getElementById( "channelSelector" );
if ( channelCombo.options.length > 0 )
{
    // we get error here as combo is not defined until later and there is no error output
    channelCombo.options[combo.selectedIndex].value;
}

var combo = <HTMLSelectElement>document.getElementById( "territoryID" );
if ( combo )
{
    // do something here
}
```

**tsconfig.json**

I am migrating large project over from JS so i am not enabling a few of the strict options just yet
```json
{
    "compileOnSave": true,
    "compilerOptions": {
        "emitBOM": true,
        "alwaysStrict": false,
        "noEmitOnError": false,
        "removeComments": false,
        "sourceMap": true,
        "target": "es3",
        "outDir": "./Scripts/"
    },
    "include": [
        "**/src/*.ts"
    ]
}
```

**Expected behavior:**
Should report a "Variable 'combo' is used before assigned" error on following line but does not:
```ts
channelCombo.options[combo.selectedIndex].value;
```

**Actual behavior:**
No error reported on undeclared usage of "combo" on line below:
```ts
channelCombo.options[combo.selectedIndex].value;
```

This error does correctly report issues in many other places, but not in this instance where the variable is mistakenly used through a typo because its declared later in the file.

**Comment by mhegazy** [comment]
the error should be reported under `"strictNullChecks": true`

**Comment by coderangerdan** [comment]
It shouldn't need to be though, I don't want to yet enable that option as I will have another 3000 errors to fix which I will do later but is unnecessary now.

The point of this issue report is that if I remove the later declaration it correctly reports combo as an undeclared variable.

This seems clearly to be a bug where later declaration is removing the error.

**Comment by mhegazy** [comment]
These errors related to how the type system sees the types of variables. with `--srtictNullChecks` on `undefined` is a type that the type system tracks. but with that flag off, it is considered to be part of the domain of all types. the feature is under a flag to avoid breaking changes to existing user code.

**Comment by coderangerdan** [comment]
Sorry, but dont know if i am missing something or you are based on my poor explanation.

The error reporting is inconsistent based on whether the variable is defined 'after' its usage. I know it reports a NULLness error if the flag is set but that is not really the error that should be reported.

This **incorrectly DOES NOT** report an error on "combo.selectedIndex":
```
var channelCombo = <HTMLSelectElement>document.getElementById( "channelSelector" );
if ( channelCombo.options.length > 0 )
{
    // we get Javascript exception here as combo is not defined until later and there is no error output
    channelCombo.options[combo.selectedIndex].value;
}

var combo = <HTMLSelectElement>document.getElementById( "territoryID" );
if ( combo )
{
    // do something here
}
```

This **correctly DOES** report a "Variable 'combo' is used before assigned" error on combo.selectedIndex":
```
var channelCombo = <HTMLSelectElement>document.getElementById( "channelSelector" );
if ( channelCombo.options.length > 0 )
{
    // we correctly get compilation error here as combo is not defined until later
    channelCombo.options[combo.selectedIndex].value;
}

// rest of code
```

**Comment by mhegazy** [comment]
> This correctly DOES report a "Variable 'combo' is used before assigned" error on combo.selectedIndex":

what is the error you get? where is `combo` defined?

**Comment by coderangerdan** [comment]
It's not defined, hence the error. It was a typo and it correctly reports usage before definition.

If you define variable AFTER its usage, the compilation error goes away when it shouldn't.

**Comment by mhegazy** [comment]
if it is not defined anywhere, then it is an error, since the compiler does not know what the name means.

if it is defined, but not initialized; then its type is `undefined` until the first initialization. this tracking is only available under `--strictNullChecks`.

**Comment by coderangerdan** [comment]
I just realised the error was not reported in the way I expected because of 'hoisting'; doh! Sorry for wasting your time.
